sdk/include/opentelemetry/sdk/resource/resource_detector.h

```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/sdk/resource/resource.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace resource
+{
+
+class ResourceDetector
+{
+public:
+  virtual std::shared_ptr<opentelemetry::sdk::resource::Resource> Detect() = 0;
```

Comment:
Same nit:  Is there a reason to return a shared_ptr here?  AFAIK Resource detection happens once and generally isn't passing pointers between different memory owners.

Reply:
As mentioned in earlier comment - As it is possible to share resources across multiple spans, it would be better to use shared_ptr and avoid un-necessary copying.

Reply:
I'd like to avoid the counting of shared_ptr references if we can just use the attached TraceProvider.

I.e. shared_ptr has a cost, and I think we don't need it in this scenario.  If we still want to use `shared_ptr` within the SDK to associate resource from `*Provider` to `Span`, that's fine but a different PR.

I think without seeing the other half of the change it's hard to fully comment on merits of shared_ptr.

Reply:
Yes, I did revisit again, and probably we mayn't need smart pointer here ( including unique_ptr). Let me add other half of implementation in this PR, so that it would be easy to comprehend and review.

Reply:
Changes are done as advised ( refer above comments ).